go/marshal/decode.go

```diff
@@ -217,3 +229,75 @@ func nomsValueDecoder(v types.Value, rv reflect.Value) {
 	}
 	rv.Set(reflect.ValueOf(v))
 }
+
+func sliceDecoder(t reflect.Type) decoderFunc {
+	decoderCache.RLock()
+	d := decoderCache.m[t]
+	decoderCache.RUnlock()
+	if d != nil {
+		return d
+	}
+
+	var decoder decoderFunc
+
+	d = func(v types.Value, rv reflect.Value) {
+		list := v.(types.List)
+		if list.Len() == 0 {
+			rv.Set(reflect.MakeSlice(t, 0, 0))
```

Comment:
I'm following encoding/json here.

One reason to not do 1 is that it would allocate more memory... another reason is that it is not consistent with how structs are filled in. We could change structs to always allocate new memory too if we wanted but it seems useful to marshal some values onto an existing struct.

```go
s := struct {
  X int `noms:"-"`
  Y int
  Z int
} {
  X: 1,
}
Unmarshal(types.NewStruct("", types.StructData{
  "Y": types.Number(2),
  "Z": types.Number(3),
}), &s)
```